transport/http2_client.go

```diff
@@ -236,19 +238,27 @@ func (t *http2Client) NewStream(ctx context.Context, callHdr *CallHdr) (_ *Strea
 			authData[k] = v
 		}
 	}
-	if _, err := wait(ctx, t.shutdownChan, t.writableChan); err != nil {
-		return nil, err
-	}
 	t.mu.Lock()
 	if t.state != reachable {
 		t.mu.Unlock()
 		return nil, ErrConnClosing
 	}
-	if uint32(len(t.activeStreams)) >= t.maxStreams {
-		t.mu.Unlock()
-		t.writableChan <- 0
-		return nil, StreamErrorf(codes.Unavailable, "transport: failed to create new stream because the limit has been reached.")
+	if t.streamsQuota != nil {
+		q, err := wait(ctx, t.shutdownChan, t.streamsQuota.acquire())
```

Comment:
On Wed, May 27, 2015 at 5:40 PM, Bryan C. Mills <notifications@github.com>
wrote:

> In transport/http2_client.go
> <https://github.com/grpc/grpc-go/pull/202#discussion_r31194751>:
>
> >  	t.mu.Lock()
> >  	if t.state != reachable {
> >  		t.mu.Unlock()
> >  		return nil, ErrConnClosing
> >  	}
> > -	if uint32(len(t.activeStreams)) >= t.maxStreams {
> > -		t.mu.Unlock()
> > -		t.writableChan <- 0
> > -		return nil, StreamErrorf(codes.Unavailable, "transport: failed to create new stream because the limit has been reached.")
> > +	if t.streamsQuota != nil {
> > +		q, err := wait(ctx, t.shutdownChan, t.streamsQuota.acquire())
>
> This call to wait occurs with t.mu still locked - is that safe?
>
fixing ...

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/202/files#r31194751>.
>